### ISSUE TITLE
Walk through

### ISSUE THREAD

**Issue by redheadedrod** [vivisect/vivisect#120]
Can someone walk me through how to use this program?

And maybe have some example code to use with this?

Looks like I will be doing some work extending this software this summer but I want to demo it in my Architecture class to the other students for my end of semester presentation. I have briefly looked through the software but if someone could walk me through a basic run of the program that would save me a LOT of time and get me on the ground running for this summer. I don't really see much documentation on using it so a little help would go a long ways. 

Thank you!

**Comment by williballenthin** [comment]
There's a lot in vivisect, and sample code is spread all over the place. Here's a few quick points that definitely do not do the tool justice.

show vivisect analysis results and ui: `vivbin -A /usr/bin/ls`
debug with vivisect vdb: `vdb -Q /usr/bin/ls`

load a workspace:

    vw = vivisect.VivWorkspace()
    vw.verbose = verbose
    vw.config.viv.parsers.pe.nx = True
    vw.loadFromFile("/usr/bin/ls")
    vw.analyze()
    vw.saveWorkspace()

disassemble an instruction:

    o = vw.parseOpcode(0x401000)
    print(o)
    print(dir(o))

list functions:

    vw.getFunctions()

get cross references:

    vw.getXrefsTo(0x401000)

get basic blocks for function:

    vw.getFunctionBlocks(0x401000)


There's hundreds of functions and thousands of things you can easily do, so this is the smallest taste of its functionality. But hopefully it gives you ideas for some things you can include during your demo.

**Comment by cmaruti** [comment]
I have collected a lot of pointers over the years to various vivisect materials. Here you go:
- https://www.singlehop.com/blog/binary-vivisection-part-1/
- https://www.singlehop.com/blog/binary-vivisection-part-2/
- https://www.singlehop.com/blog/binary-vivisection-part-3/
- http://nest.unm.edu/files/5514/1254/9114/Malware_Analysis_with_Vivsect.pdf
- https://www.fireeye.com/blog/threat-research/2013/02/custom-vdb-debugger-exploit-analysis.html
- https://github.com/fireeye/flare-dbg/
- http://shellsherpa.nl/stealing-keepasswords-with-vivisect
- http://dvlabs.tippingpoint.com/blog/2012/04/02/mindshare-vtrace-input-tracking

A bit outdated but still a good references:
- https://www.defcon.org/images/defcon-16/dc16-presentations/defcon-16-atlas.pdf
- https://github.com/pdasilva/vtrace_scripts

**Comment by redheadedrod** [comment]
Thanks for the pointers. I will try working with this this weekend and go from there. I am going to try doing a demo of using this program for my Architecture class  so I have a couple weeks to learn how to use it. 

I also received some other pointers offline so hopefully I have enough to get started playing with it.

Rodney

**Comment by redheadedrod** [comment]
While trying these there seems to be a bug/issue that has come up. 
With the part-1 the file reverseme.exe is converted to a .viv file and then opened up in the gui. 
The instructions say to click on the "strings" tab which has nothing in it. I have tried this with other programs and I am
getting the same result. I have tried this on a machine running Windows 10, Windows 8.1 and under Ubuntu 15.10 and they all show the same issue so I am under the impression it is a bug at this point in time.

Or am I missing something?
I have been able to do some other stuff and look through stuff as well but I just don't see the strings.
I have also tried the atlas fork as well as the main fork for this.

Rodney

Date: Wed, 30 Mar 2016 08:42:15 -0700

**Comment by cmaruti** [comment]
Hi @redheadedrod 
first of all did you load a layout? (Use _View->Layouts->Load and select vivisect/qt/default.lyt_).
I just tested the binary referenced in the tutorial and I can see the strings pane correctly filled up...

![vivi-reverseme](https://cloud.githubusercontent.com/assets/2283072/14420500/b42446fa-ffcc-11e5-8213-ca9331d9b3b6.png)

**Comment by redheadedrod** [comment]
Yes I am using the default load out. 
I downloaded the latest vivisect from github by downloading a zip and copied this to its own directory as late as 4/8.

I am using anaconda under windows 10 and loaded the pyqt class.
Using ipython which is using the 2.7 python kernel. 
Everything else I try is working so far but not the strings portion. Ive loaded all three of the programs mentioned in the first 3 tutorials but I am not seeing strings in any of them.

Ive also tried it under windows 8.1 and Ubuntu 15.10. The ubuntu I installed ipython via anaconda.

I wonder if I am missing something in the folder that might be preventing it from working?

Did you try it with the current github version?

I have also had issues when I maximize the screen that shows the graph. When I close it out my instance of vivisect crashes and I have to restart it.  So I have learned not to maximize the graph. 

Thanks for your help! I really want to get down to the nitty gritty with this program in the next week or so since I hope to demo using this program end of next week as my presentation in my architecture class. 

I look forward to working with this framework this summer and hopefully I will be productive in extending it in whatever manner my mentor has in mind for me. I do appreciate any help you or anyone else provides as it accelerates my understanding and hopefully will help me be productive that much sooner.

**Comment by cmaruti** [comment]
Just checked out the last master and I had the same problem reported by @redheadedrod cc: @atlas0fd00m @invisig0th

**Comment by atlas0fd00m** [comment]
i've been noticing a serious degradation of string identification recently.  so much so that i've written a temporary string-scraping script for some firmware i'm tearing apart right now.  we need to look into it better.  i believe i saw that only function analysis looks for strings in xrefs from within the functions.  ideally that would be enough, but with string arrays and such, this seems insufficient.

**Comment by redheadedrod** [comment]
It seems to be a sporatic thing. I have tried running it on my laptop yesterday under Windows 10 and something must have been updated because now the strings are working but the layout system is touchy but just on my laptop. If I load a new layout instead of replacing the current layout it just takes some portion of the screen and combines it together in a weird manner. On my desktop running Win 8.1 but with same exact ipython setup I don't see the strings but the layouts work fine. I will try working with this under ubuntu a little more and see if anything changes. The lack of strings issue was happening on my desktop on a ubuntu 15.10 VM but my laptop triple boots so I will try to install it under Ubuntu 15.10 and Kali2.0 rolling and see if I have any better luck. 


Date: Mon, 11 Apr 2016 08:54:04 -0700

**Comment by atlas0fd00m** [comment]
the strings analysis should happen during Bulk Analysis (vivbin -B somebin.exe) which doesn't use the GUI.  then the generated file (somebin.exe.viv) can be loaded on either machine.  

can you describe exactly what happens?  try doing bulk analysis on both systems and loading both generated .viv files on both machines.  let us know if you can isolate anything further.  differences between systems is very odd for viv.

@

-- 
GUYTKOBVGE3DONJZGUZTIMRWIE2UCNJYGRATGMBVHE2TONSDG42TIOJUHA2DENZWGYYTKNZTGUZTANBZGQ3TMQZXGU2DSNBYGUZDOMBWGI2TONJVG4ZTIOJUHA2TENSGGVATKMZUGI3TSNKBGU4DINRTGE3DCNJYGRATMQZVIE2DGNBSG4ZTKQJVHA2UCNSDGYZDIMZUGI3TMNKBGY4TIMRWIE3DEMZSGMYTONZWGM3EINJWGZDDKQJVG4ZTKN2BGYYTKNZTHE3TKNBZGQ4DKMRXGY2DSNBXGQ3DMQJWGE2DONSDGZBTMNBWIQ2TKNRXGU4TKMZUGI3UCNRUGU3TIRJWIE2UCNJYGRCTOQJVIE3EKNJWG4ZTIOJUG42TMMZUGYZTINZXHA3TMNRRGU4DKMRWHA3DINBXGZBTONRWGI3DSNBSG44TKQJVG42DMNSBGYYTINZVGY3UCNBZGQ4DKMRWIY2UCNJTGQZDMQZVIE2DONRUGZBTIOJUG4ZTSNSEGQ4TINZWIM3TKNRTGMZDINRXGU3DCNJYGUZDGNJUIM3DSNBSGU2TMMJUG42TKNRXGYZTGMZUGI3TANRTGZCDMQZTGA2DSNBXGMYTMQZWGI2DQNJSG5ATIOJUHA3DINZQGY2DINZWG43DONRSGU3TINRWII3DENSEGU3DOQJWGM3TSNZXGY3TMNBTGI2TKNRXGY2DINZVGY3TINRTGQ3TGOJXHE2TSNJYGRATOMBWGI2DQNSCGY3TMMRUG4ZTSNZWGYZTGMRVGU3DONJZGU3TOOBXGM2DSNBXG42DOMBWGI3EINJRGY3TMMRTGI2TSNRXGYZTMRBUGYZTANRRGU3TGOJXGU2TSNJXG44DOMBWGQ2DQNSCGY3TKOJVG4ZTKNSCGQ4TIOBWGQ3EGNBZGQ3TKNRXGU3DINBXGU3DOOJUHE2DONBVGY3TMMZTGM2TENRYGY2DINZVGU3DONRSGMZDKOJWG43DCNJXG44DOMZWGQ2TOMZRG4YDMMRWIQ2DMMZQGYYTKNZTHE3TKNCDGY4TIMJXGQ2DSNBYGVATMQRVGE2DQNBSGZDDMMZWIQ2DMNSBGYYTOOJTGU3TMNRTGZCDMMZUII======

**Comment by redheadedrod** [comment]
How I am doing this is running vivbin -B reverseme.exe which produces a reverseme.exe.viv file.

I then load this by using the command vivbin reverseme.exe.viv

I have tried this both in windows and linux and get the same result. This is a simple windows program that expects a key and kicks out with "evaluation period has ended". It is special made for reverse engineering to remove the "evaluation period has ended" warning. When you bypass the key it says something like "You completed it!"

While I can see strings in the .DATA segment I am seeing absolutely nothing in the Strings tab...
The strings in the .DATA segment show references in the listing so the program knows they are there.
And they also end with a 00 so it should be easy for the program to pick up those strings.

I get a little different response when running ELF binaries.

I compiled conditional.c from one of the tutorial links and loaded it via using the same command as before:
vivbin -B conditional
Then run the GUI via:
vivbin conditional.viv

With this I DO get some entries in the strings tab but they seem to be references to libraries installed by the program and I see a puts\x00 and printf\x00  Otherwise it appears to just be listing the libraries it is using and some versioning related to the libraries.

I do seem to be getting the same results with the strings tab regardless of where I am going and is very repeatable. I am not familiar enough (yet) with ELF format to know where to find strings.

Update on the layouts issue I was having too...
When loading a new layout the layout is laid on top of the current layout. In order to properly load a new layout I have to remove all of the windows until I just have the interactive window. Shouldn't the program clear the current layout before loading a new layout over the top of the screen? It really looks weird if you load the default layout over the top of the default layout.

Note I am a new user to this program and have not touched upon the CLI much yet. I have been trying to understand the program as much as I can for my presentation next week.
